src/instrument.ts

```diff
@@ -0,0 +1,249 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  SEMATTRS_DB_NAME,
+  SEMATTRS_DB_STATEMENT,
+  SEMATTRS_DB_SYSTEM,
+  SEMATTRS_DB_SQL_TABLE,
+} from '@opentelemetry/semantic-conventions';
+
+import {
+  Span,
+  SpanStatusCode,
+  trace,
+  INVALID_SPAN_CONTEXT,
+  SpanAttributes,
+  TimeInput,
+  TracerProvider,
+  Link,
+  Exception,
+  SpanContext,
+  SpanStatus,
+  SpanKind,
+} from '@opentelemetry/api';
+
+const optedInPII: boolean =
+  process.env.SPANNER_ENABLE_EXTENDED_TRACING === 'true';
+
+interface SQLStatement {
+  sql: string;
+}
+
+interface observabilityOptions {
+  tracerProvider: TracerProvider;
+  enableExtendedTracing: boolean;
+}
+
+export type {observabilityOptions as ObservabilityOptions};
+
+const TRACER_NAME = 'cloud.google.com/nodejs/spanner';
+const TRACER_VERSION = '7.14.0'; // Manually hard coded, TODO: remove
+
+/**
+ * getTracer fetches the tracer from the provided tracerProvider.
+ * @param {TracerProvider} [tracerProvider] -- optional custom tracer provider
+ * to use for fetching the tracer. If not provided, the global provider will be used.
+ *
+ * @returns {Tracer} The tracer instance associated with the provided or global provider.
+ */
+export function getTracer(tracerProvider?: TracerProvider) {
+  if (tracerProvider) {
+    return tracerProvider.getTracer(TRACER_NAME, TRACER_VERSION);
+  }
+  // Otherwise use the global tracer.
+  return trace.getTracer(TRACER_NAME, TRACER_VERSION);
+}
+
+interface traceConfig {
+  sql?: string | SQLStatement;
+  tableName?: string;
+  dbName?: string;
+  opts?: observabilityOptions;
+  that?: Object;
+}
+
+const SPAN_NAMESPACE_PREFIX = 'cloud.google.com/nodejs/spanner';
+export {SPAN_NAMESPACE_PREFIX};
+
+/**
+ * startTrace begins an active span in the current active context
+ * and passes it back to the set callback function. Each span will
+ * be prefixed with "cloud.google.com/nodejs/spanner". It is the
+ * responsibility of the caller to invoke [span.end] when finished tracing.
+ *
+ * @returns {Span} -- The created span.
+ */
+export function startTrace<T>(
+  spanNameSuffix: string,
+  config: traceConfig | undefined,
+  cb: (span: Span) => T
+): T {
+  const origConfig = config;
+  if (!config) {
+    config = {} as traceConfig;
+  }
+
+  return getTracer(config.opts?.tracerProvider).startActiveSpan(
+    SPAN_NAMESPACE_PREFIX + '/' + spanNameSuffix,
+    {kind: SpanKind.CLIENT},
+    span => {
+      span.setAttribute(SEMATTRS_DB_SYSTEM, 'spanner');
+
+      if (config.tableName) {
+        span.setAttribute(SEMATTRS_DB_SQL_TABLE, config.tableName);
+      }
+      if (config.dbName) {
+        span.setAttribute(SEMATTRS_DB_NAME, config.dbName);
+      }
+
+      // If they optedInPII but opts.enableExtendedTracing=false, reject it.
```

Comment:
@odeke-em There is a change which we discussed for env variables. 

ExtendedTracing should be enabled if either env variable is enabled or opts.enableExtendedTracing is enabled

Reply:
Yes it is enabled for either, but this case handles when the environment variable is true but the library caller explicitly passes in `{opts:{enableExtendedTracing:false}}`

Reply:
No even in this case we should make `enableExtendedTracing ` as true. So consider customer has set  `{opts:{enableExtendedTracing:false}}` but during some issue, they dont want to redeploy their aplication just set the environment variable and restart . 

Sorry I did not inform this earlier, this is a discussion we had yesterday.

Reply:
@surbhigarg92 alright, thank you, I've simplified it even more.